pkg/demoinfocs/datatables.go

```diff
@@ -632,12 +755,45 @@ func (p *parser) bindGrenadeProjectiles(entity st.Entity) {
 
 		p.gameEventHandler.addThrownGrenade(proj.Thrower, proj.WeaponInstance)
 
+		if p.demoInfoProvider.IsSource2() && wep == common.EqFlash {
+			p.gameState.flyingFlashbangs = append(p.gameState.flyingFlashbangs, &FlyingFlashbang{
+				projectile:       proj,
+				flashedEntityIDs: []int{},
+			})
+		}
+
+		if p.isSource2() {
+			p.eventDispatcher.Dispatch(events.WeaponFire{
+				Shooter: proj.Owner,
+				Weapon:  proj.WeaponInstance,
+			})
+		}
+
 		p.eventDispatcher.Dispatch(events.GrenadeProjectileThrow{
 			Projectile: proj,
 		})
 	})
 
 	entity.OnDestroy(func() {
```

Comment:
can we verify this? sometimes entities are not destroyed when they are removed from the scene.

CS:GO re-uses entities of the same type to improve memory allocation

Reply:
I added print calls in the `OnCreateFinished` and `OnDestroy` callbacks and used the following code to verify:

```go
package main

import (
	"errors"
	"fmt"
	"os"
	"time"

	demoinfocs "github.com/markus-wa/demoinfocs-golang/v3/pkg/demoinfocs"
	"github.com/markus-wa/demoinfocs-golang/v3/pkg/demoinfocs/common"
	"github.com/markus-wa/demoinfocs-golang/v3/pkg/demoinfocs/events"
)

func main() {
	var err error
	f, err := os.Open("./faze-vs-legends-m2-nuke.dem")
	checkError(err)

	defer f.Close()

	p := demoinfocs.NewParser(f)
	defer p.Close()

	_, err = p.ParseHeader()
	checkError(err)

	p.RegisterEventHandler(func(e events.WeaponFire) {
		if e.Weapon.Type == common.EqFlash {
			fmt.Println(p.GameState().IngameTick(), "weapon_fire", e.Shooter.Name, e.Weapon.Type)
		}
	})

	p.RegisterEventHandler(func(e events.FlashExplode) {
		fmt.Println(p.GameState().IngameTick(), "flash_explode", e.Thrower.Name)
	})

	now := time.Now()
	err = p.ParseToEnd()
	if errors.Is(err, demoinfocs.ErrUnexpectedEndOfDemo) {
		fmt.Println("Parsing done with unexpected EOF")
	} else {
		checkError(err)
	}
	fmt.Println(time.Since(now))
}

func checkError(err error) {
	if err != nil {
		panic(err)
	}
}
```

<details>
<summary>Output example</summary>

7911 weapon_fire twistzz Flashbang
7911 created 451
8007 weapon_fire twistzz Flashbang
8007 created 95
8016 destroy 451
8016 flash_explode twistzz
8112 destroy 95
8112 flash_explode twistzz
12231 weapon_fire broky Flashbang
12231 created 150
12336 destroy 150
12336 flash_explode broky
14969 weapon_fire rain Flashbang
14969 created 496
15047 weapon_fire KARRiGAN Flashbang
15047 created 89
15074 destroy 496
15074 flash_explode rain
15152 destroy 89
15152 flash_explode KARRiGAN
16127 weapon_fire twistzz Flashbang
16127 created 427
16232 destroy 427
16232 flash_explode twistzz
16535 weapon_fire broky Flashbang
16535 created 423
16640 destroy 423
16640 flash_explode broky
16862 weapon_fire twistzz Flashbang
16862 created 405
16967 destroy 405
16967 flash_explode twistzz
19547 weapon_fire NBK- Flashbang
19547 created 346
19652 destroy 346
19652 flash_explode NBK-
19820 weapon_fire broky Flashbang
19820 created 111
19871 weapon_fire twistzz Flashbang
19871 created 99
19925 destroy 111
19925 flash_explode broky
19961 weapon_fire twistzz Flashbang
19961 created 185
19976 destroy 99
19976 flash_explode twistzz
20054 weapon_fire KARRiGAN Flashbang
20054 created 186
20066 destroy 185
20066 flash_explode twistzz
20159 destroy 186
20159 flash_explode KARRiGAN
24924 weapon_fire ropz Flashbang
24924 created 269
24927 weapon_fire NBK- Flashbang
24927 created 271
25029 destroy 269
25029 flash_explode ropz
25032 destroy 271
25032 flash_explode NBK-
25542 weapon_fire broky Flashbang
25542 created 294
25635 weapon_fire broky Flashbang
25635 created 303
25647 destroy 294
25647 flash_explode broky
25740 destroy 303
25740 flash_explode broky
28294 weapon_fire twistzz Flashbang
28294 created 384
28396 weapon_fire twistzz Flashbang
28396 created 218
28399 destroy 384
28399 flash_explode twistzz
28501 destroy 218
28501 flash_explode twistzz
32757 weapon_fire rain Flashbang
32757 created 126
32829 weapon_fire twistzz Flashbang
32829 created 494
32862 destroy 126
32862 flash_explode rain
32910 weapon_fire neo Flashbang
32910 created 184
32934 destroy 494
32934 flash_explode twistzz
33000 weapon_fire NBK- Flashbang
33000 created 174
33015 destroy 184
33015 flash_explode neo
33057 weapon_fire twistzz Flashbang
33057 created 398
33105 destroy 174
33105 flash_explode NBK-
33162 destroy 398
33162 flash_explode twistzz
38553 weapon_fire KARRiGAN Flashbang
38553 created 106
38556 weapon_fire NBK- Flashbang
38556 created 70
38601 weapon_fire ropz Flashbang
38601 created 450
38658 destroy 106
38658 flash_explode KARRiGAN
38661 destroy 70
38661 flash_explode NBK-
38706 destroy 450
38706 flash_explode ropz
39912 weapon_fire neo Flashbang
39912 created 510
40017 destroy 510
40017 flash_explode neo
42599 weapon_fire NBK- Flashbang
42599 created 171
42611 weapon_fire KARRiGAN Flashbang
42611 created 283
42704 destroy 171
42704 flash_explode NBK-
42716 destroy 283
42716 flash_explode KARRiGAN
44369 weapon_fire twistzz Flashbang
44369 created 377
44474 destroy 377
44474 flash_explode twistzz
46568 weapon_fire neo Flashbang
46568 created 314
46673 destroy 314
46673 flash_explode neo
47987 weapon_fire broky Flashbang
47987 created 375
48092 destroy 375
48092 flash_explode broky
50694 weapon_fire twistzz Flashbang
50694 created 312
50787 weapon_fire twistzz Flashbang
50787 created 258
50799 destroy 312
50799 flash_explode twistzz
50892 destroy 258
50892 flash_explode twistzz
50898 weapon_fire Ladi Flashbang
50898 created 148
51003 destroy 148
51003 flash_explode Ladi
51228 weapon_fire ropz Flashbang
51228 created 197
51333 destroy 197
51333 flash_explode ropz
52776 weapon_fire rain Flashbang
52776 created 153
52881 destroy 153
52881 flash_explode rain
56416 weapon_fire broky Flashbang
56416 created 471
56521 destroy 471
56521 flash_explode broky
56809 weapon_fire KARRiGAN Flashbang
56809 created 436
56914 destroy 436
56914 flash_explode KARRiGAN
59712 weapon_fire rain Flashbang
59712 created 71
59817 destroy 71
59817 flash_explode rain
60471 weapon_fire Ladi Flashbang
60471 created 76
60576 destroy 76
60576 flash_explode Ladi
60957 weapon_fire neo Flashbang
60957 created 130
61062 destroy 130
61062 flash_explode neo
62019 weapon_fire broky Flashbang
62019 created 304
62124 destroy 304
62124 flash_explode broky
66826 weapon_fire twistzz Flashbang
66826 created 239
66850 weapon_fire NBK- Flashbang
66850 created 281
66913 weapon_fire broky Flashbang
66913 created 171
66919 weapon_fire twistzz Flashbang
66919 created 264
66931 destroy 239
66931 flash_explode twistzz
66955 destroy 281
66955 flash_explode NBK-
67018 destroy 171
67018 flash_explode broky
67024 destroy 264
67024 flash_explode twistzz
67042 weapon_fire rain Flashbang
67042 created 204
67102 weapon_fire KARRiGAN Flashbang
67102 created 187
67147 destroy 204
67147 flash_explode rain
67207 destroy 187
67207 flash_explode KARRiGAN
67408 weapon_fire papito Flashbang
67408 created 210
67513 destroy 210
67513 flash_explode papito
71716 weapon_fire rain Flashbang
71716 created 151
71821 destroy 151
71821 flash_explode rain
71830 weapon_fire NBK- Flashbang
71830 created 345
71935 destroy 345
71935 flash_explode NBK-
72327 weapon_fire broky Flashbang
72327 created 480
72432 destroy 480
72432 flash_explode broky
72873 weapon_fire broky Flashbang
72873 created 413
72978 destroy 413
72978 flash_explode broky
73341 weapon_fire Ladi Flashbang
73341 created 256
73446 destroy 256
73446 flash_explode Ladi
75725 weapon_fire NBK- Flashbang
75725 created 253
75830 destroy 253
75830 flash_explode NBK-
76163 weapon_fire KARRiGAN Flashbang
76163 created 268
76268 destroy 268
76268 flash_explode KARRiGAN
80903 weapon_fire NBK- Flashbang
80903 created 449
81008 destroy 449
81008 flash_explode NBK-
88035 weapon_fire papito Flashbang
88035 created 308
88128 weapon_fire papito Flashbang
88128 created 282
88140 destroy 308
88140 flash_explode papito
88233 destroy 282
88233 flash_explode papito
94184 weapon_fire broky Flashbang
94184 created 376
94289 destroy 376
94289 flash_explode broky
94463 weapon_fire NBK- Flashbang
94463 created 313
94568 destroy 313
94568 flash_explode NBK-
95129 weapon_fire allengOlf~ Flashbang
95129 created 317
95234 destroy 317
95234 flash_explode allengOlf~
97007 weapon_fire Ladi Flashbang
97007 created 328
97112 destroy 328
97112 flash_explode Ladi
103623 weapon_fire neo Flashbang
103623 created 431
103728 destroy 431
103728 flash_explode neo
104040 weapon_fire Ladi Flashbang
104040 created 322
104115 weapon_fire papito Flashbang
104115 created 250
104145 destroy 322
104145 flash_explode Ladi
104220 destroy 250
104220 flash_explode papito
104868 weapon_fire NBK- Flashbang
104868 created 65
104973 destroy 65
104973 flash_explode NBK-
105240 weapon_fire allengOlf~ Flashbang
105240 created 74
105345 destroy 74
105345 flash_explode allengOlf~
105738 weapon_fire KARRiGAN Flashbang
105738 created 266
105843 destroy 266
105843 flash_explode KARRiGAN
108930 weapon_fire rain Flashbang
108930 created 196
109035 destroy 196
109035 flash_explode rain
109098 weapon_fire twistzz Flashbang
109098 created 251
109203 destroy 251
109203 flash_explode twistzz
113447 weapon_fire broky Flashbang
113447 created 114
113552 destroy 114
113552 flash_explode broky
113657 weapon_fire twistzz Flashbang
113657 created 243
113762 destroy 243
113762 flash_explode twistzz
114530 weapon_fire broky Flashbang
114530 created 104
114635 destroy 104
114635 flash_explode broky
114941 weapon_fire rain Flashbang
114941 created 386
115046 destroy 386
115046 flash_explode rain
118473 weapon_fire twistzz Flashbang
118473 created 417
118578 destroy 417
118578 flash_explode twistzz
119106 weapon_fire broky Flashbang
119106 created 302
119211 destroy 302
119211 flash_explode broky
121414 weapon_fire Ladi Flashbang
121414 created 116
121519 destroy 116
121519 flash_explode Ladi
121579 weapon_fire neo Flashbang
121579 created 171
121684 destroy 171
121684 flash_explode neo
121753 weapon_fire allengOlf~ Flashbang
121753 created 200
121858 destroy 200
121858 flash_explode allengOlf~
123811 weapon_fire allengOlf~ Flashbang
123811 created 214
123916 destroy 214
123916 flash_explode allengOlf~
126385 weapon_fire NBK- Flashbang
126385 created 381
126490 destroy 381
126490 flash_explode NBK-
128509 weapon_fire NBK- Flashbang
128509 created 259
128614 destroy 259
128614 flash_explode NBK-
129127 weapon_fire twistzz Flashbang
129127 created 263
129232 destroy 263
129232 flash_explode twistzz
130264 weapon_fire KARRiGAN Flashbang
130264 created 272
130369 destroy 272
130369 flash_explode KARRiGAN
132619 weapon_fire neo Flashbang
132619 created 299
132661 weapon_fire allengOlf~ Flashbang
132661 created 425
132724 destroy 299
132724 flash_explode neo
132765 destroy 425
132765 flash_explode allengOlf~
133164 weapon_fire papito Flashbang
133164 created 339
133270 destroy 339
133270 flash_explode papito
137175 weapon_fire Ladi Flashbang
137175 created 94
137244 weapon_fire ropz Flashbang
137244 created 99
137271 weapon_fire Ladi Flashbang
137271 created 74
137280 destroy 94
137280 flash_explode Ladi
137349 destroy 99
137349 flash_explode ropz
137376 destroy 74
137376 flash_explode Ladi
137721 weapon_fire broky Flashbang
137721 created 266
137826 destroy 266
137826 flash_explode broky
139680 weapon_fire KARRiGAN Flashbang
139680 created 386
139785 destroy 386
139785 flash_explode KARRiGAN
142664 weapon_fire Ladi Flashbang
142664 created 84
142769 destroy 84
142769 flash_explode Ladi
142805 weapon_fire ropz Flashbang
142805 created 324
142910 destroy 324
142910 flash_explode ropz
143462 weapon_fire allengOlf~ Flashbang
143462 created 209
143567 destroy 209
143567 flash_explode allengOlf~
143777 weapon_fire papito Flashbang
143777 created 211
143882 destroy 211
143882 flash_explode papito
144083 weapon_fire twistzz Flashbang
144083 created 144
144188 destroy 144
144188 flash_explode twistzz
144368 weapon_fire neo Flashbang
144368 created 327
144473 destroy 327
144473 flash_explode neo
147983 weapon_fire Ladi Flashbang
147983 created 407
148088 destroy 407
148088 flash_explode Ladi
148157 weapon_fire broky Flashbang
148157 created 472
148262 destroy 472
148262 flash_explode broky
148307 weapon_fire rain Flashbang
148307 created 275
148412 destroy 275
148412 flash_explode rain
150086 weapon_fire neo Flashbang
150086 created 509
150191 destroy 509
150191 flash_explode neo
152168 weapon_fire rain Flashbang
152168 created 273
152270 weapon_fire ropz Flashbang
152270 created 494
152273 destroy 273
152273 flash_explode rain
152375 destroy 494
152375 flash_explode ropz
152951 weapon_fire KARRiGAN Flashbang
152951 created 335
153056 destroy 335
153056 flash_explode KARRiGAN
161512 weapon_fire NBK- Flashbang
161512 created 440
161560 weapon_fire twistzz Flashbang
161560 created 320
161611 weapon_fire NBK- Flashbang
161611 created 111
161617 destroy 440
161617 flash_explode NBK-
161665 destroy 320
161665 flash_explode twistzz
161716 destroy 111
161716 flash_explode NBK-
161965 weapon_fire neo Flashbang
161965 created 371
162070 destroy 371
162070 flash_explode neo
162697 weapon_fire broky Flashbang
162697 created 235
162802 destroy 235
162802 flash_explode broky
164335 weapon_fire broky Flashbang
164335 created 218
164440 destroy 218
164440 flash_explode broky
167531 weapon_fire ropz Flashbang
167531 created 155
167579 weapon_fire broky Flashbang
167579 created 444
167636 destroy 155
167636 flash_explode ropz
167684 destroy 444
167684 flash_explode broky
167783 weapon_fire neo Flashbang
167783 created 188
167888 destroy 188
167888 flash_explode neo
168011 weapon_fire twistzz Flashbang
168011 created 338
168114 destroy 338
168114 flash_explode twistzz
169137 weapon_fire KARRiGAN Flashbang
169137 created 450
169242 destroy 450
169242 flash_explode KARRiGAN
172398 weapon_fire Ladi Flashbang
172398 created 481
172503 destroy 481
172503 flash_explode Ladi
174888 weapon_fire neo Flashbang
174888 created 269
174993 destroy 269
174993 flash_explode neo
176196 weapon_fire rain Flashbang
176196 created 458
176217 weapon_fire KARRiGAN Flashbang
176217 created 234
176277 weapon_fire neo Flashbang
176277 created 239
176301 destroy 458
176301 flash_explode rain
176322 destroy 234
176322 flash_explode KARRiGAN
176382 destroy 239
176382 flash_explode neo
179666 weapon_fire NBK- Flashbang
179666 created 115
179771 destroy 115
179771 flash_explode NBK-
180128 weapon_fire NBK- Flashbang
180128 created 501
180233 destroy 501
180233 flash_explode NBK-
180347 weapon_fire KARRiGAN Flashbang
180347 created 391
180452 destroy 391
180452 flash_explode KARRiGAN
180959 weapon_fire twistzz Flashbang
180959 created 157
181058 weapon_fire rain Flashbang
181058 created 150
181064 destroy 157
181064 flash_explode twistzz
181163 destroy 150
181163 flash_explode rain
184080 weapon_fire broky Flashbang
184080 created 287
184185 destroy 287
184185 flash_explode broky
184788 weapon_fire twistzz Flashbang
184788 created 408
184893 destroy 408
184893 flash_explode twistzz
185010 weapon_fire KARRiGAN Flashbang
185010 created 432
185115 destroy 432
185115 flash_explode KARRiGAN
185175 weapon_fire twistzz Flashbang
185175 created 222
185280 destroy 222
185280 flash_explode twistzz
185283 weapon_fire neo Flashbang
185283 created 269
185388 destroy 269
185388 flash_explode neo
Parsing done with unexpected EOF

</details>

As you can see, flashbang entities are always created and then destroyed (tried with 3 demos).

Reply:
nice, thx!